src/exe.ml

```diff
@@ -138,6 +143,72 @@ let link_exe
       Build.dyn_paths (Build.arr (fun (modules, _) ->
         artifacts modules ~ext:ctx.ext_obj))
   in
+  let link_flags_for_requires =
+    Result.map requires ~f:(fun libs ->
```

Comment:
Could you move this code to its own module? It's looks big enough and self-contained, and in the future I hope that we'll also support a generic method for link-time code generation (https://github.com/ocaml/dune/issues/594). So for instance you can move it to `Link_time_code_gen`.